xdsl/transforms/mpi_to_llvm.py

```diff
@@ -0,0 +1,525 @@
+import dataclasses
+from abc import ABC
+from typing import TypeVar, cast
+
+from xdsl.dialects.builtin import Signedness, IntegerType
+from xdsl.ir import Operation, SSAValue, OpResult, Attribute, MLContext
+
+from xdsl.pattern_rewriter import RewritePattern, PatternRewriter, op_type_rewrite_pattern, PatternRewriteWalker, \
+    AnonymousRewritePattern, GreedyRewritePatternApplier
+from xdsl.dialects import mpi, llvm, func, memref, arith, builtin
+
+
+@dataclasses.dataclass
+class MpiLibraryInfo:
+    """
+    This object is meant to capture characteristics of a specific MPI implementations.
+
+    It holds magic values, sizes of structs, field offsets and much more.
+
+    We need these as we currently cannot load these library headers into the programs we want to lower,
+    therefore we need to generate our own external stubs and load magic values directly.
+
+    This way of doing it is inherently fragile, but we don't know of any better way.
+    We plan to include a C file that automagically extracts all this information from MPI headers.
+
+    These defaults have been chosen to work with **our** version of OpenMPI. No guarantees of portability!
+    """
+    mpi_comm_world_val: int = 0x44000000
+
+    MPI_INT: int = 0x4c000405
+    MPI_UNSIGNED: int = 0x4c000406
+    MPI_LONG: int = 0x4c000807
+    MPI_UNSIGNED_LONG: int = 0x4c000808
+    MPI_FLOAT: int = 0x4c00040a
+    MPI_DOUBLE: int = 0x4c00080b
+    MPI_UNSIGNED_CHAR: int = -1
+    MPI_UNSIGNED_SHORT: int = -1
+    MPI_UNSIGNED_LONG_LONG: int = -1
+    MPI_CHAR: int = -1
+    MPI_SHORT: int = -1
+    MPI_LONG_LONG_INT: int = -1
+
+    MPI_STATUS_IGNORE: int = 1
+
+    request_size: int = 4
+    status_size: int = 4 * 5
+    mpi_comm_size: int = 4
+
+
+_RewriteT = TypeVar('_RewriteT', bound=mpi.MPIBaseOp)
+
+
+class _MPIToLLVMRewriteBase(RewritePattern, ABC):
+    """
+    This base is used to convert a pure rewrite function `lower` into the impure
+    `match_and_rewrite` method used by xDSL.
+
+    In order to lower that far, we require some information about the targeted MPI library
+    (magic values, struct sizes, field offsets, etc.). This information is provided using
+    the MpiLibraryInfo class.
+    """
+
+    MPI_SYMBOL_NAMES = {
+        'mpi.init': 'MPI_Init',
+        'mpi.finalize': 'MPI_Finalize',
+        'mpi.irecv': 'MPI_Irecv',
+        'mpi.isend': 'MPI_Isend',
+        'mpi.wait': 'MPI_Wait',
+        'mpi.comm.rank': 'MPI_Comm_rank',
+        'mpi.recv': 'MPI_Recv',
+        'mpi.send': 'MPI_Send'
+    }
+    """
+    Translation table for mpi operation names to their MPI library function names
+    """
+
+    info: MpiLibraryInfo
+    """
+    This object carries information about the targeted MPI libraray
+    """
+
+    def __init__(self, info: MpiLibraryInfo):
```

Comment:
Can we make this class a dataclass, and remove this?

Reply:
I'm not sure, then we'd have to have all the rewrites also subclass from dataclass, right? I think this is fine to have?

Reply:
Okay, done!